lxd/storage/drivers/driver_btrfs.go

```
@@ -0,0 +1,1514 @@
+				}
+				if len(subvols) > 0 {
+					return fmt.Errorf("Requested btrfs subvolume exists but is not empty")
+				}
+			} else {
+				cleanSource := filepath.Clean(source)
+				lxdDir := shared.VarPath()
+
+				if shared.PathExists(source) && !isOnBtrfs(source) {
+					return fmt.Errorf("Existing path is neither a btrfs subvolume nor does it reside on a btrfs filesystem")
+				} else if strings.HasPrefix(cleanSource, lxdDir) {
+					if cleanSource != poolMntPoint {
+						return fmt.Errorf("btrfs subvolumes requests in LXD directory %q are only valid under %q\n(e.g. source=%s)", shared.VarPath(), shared.VarPath("storage-pools"), poolMntPoint)
+					} else if d.state.OS.BackingFS != "btrfs" {
+						return fmt.Errorf("Creation of btrfs subvolume requested but %q does not reside on btrfs filesystem", source)
+					}
+				}
+
+				err := btrfsSubvolumeCreate(source)
+				if err != nil {
+					return err
+				}
+			}
+		}
+	}
+
+	var err error
+	var devUUID string
+	mountFlags, mountOptions := ResolveMountOptions(d.getMountOptions())
+	mountFlags |= d.remount
+
+	if isBlockDev {
+		devUUID, _ = shared.LookupUUIDByBlockDevPath(source)
+		// The symlink might not have been created even with the delay
+		// we granted it above. So try to call btrfs filesystem show and
+		// parse it out. (I __hate__ this!)
+		if devUUID == "" {
+			devUUID, err = btrfsLookupFsUUID(source)
+			if err != nil {
+				return err
+			}
+		}
+		d.config["source"] = devUUID
+
+		// If the symlink in /dev/disk/by-uuid hasn't been created yet
+		// aka we only detected it by parsing btrfs filesystem show, we
+		// cannot call StoragePoolMount() since it will try to do the
+		// reverse operation. So instead we shamelessly mount using the
+		// block device path at the time of pool creation.
+		err = tryMount(source, GetPoolMountPath(d.name), "btrfs", mountFlags, mountOptions)
+	} else {
+		_, err = d.Mount()
+	}
+	if err != nil {
+		return err
+	}
```

I'd prefer if each of the two paths did its own error checking, we'd be less likely to accidentally add code afterwards, masking the errors.